### PR TITLE
feat(squash): add squash command and action

### PR DESCRIPTION
**Pull request by sileht**
Fixes MRGFY-284

Change-Id: I26d3c9bb5be07768cff182707309176880a18b6c

### CODE REVIEW

**Comment by AraHaan** [comment]
So basically the new squash action would act as the merge action but instead in the squashed state or no?

**Comment by sileht** [comment]
> So basically the new squash action would act as the merge action but instead in the squashed state or no?

No, it just squashes the pull request without merging it.

**Review by jd — CHANGES_REQUESTED** [review]

File: docs/source/actions/squash.rst [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+   :widths: 1 1 1 3
+
+   * - Key Name
+     - Value Type
+     - Default
+     - Value Description
+   * - ``message``
+     - :ref:`data type template`
+     -
+     - The message to write as a squash.
```
[the remark this answers is not in the document]
Comment (sileht) [reply docs/source/actions/squash.rst]: Me neither.

File: docs/source/actions/squash.rst [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+     - ``default``
+     - Defines what commit message to use when merging using the ``squash`` or
+       ``merge`` method. Possible values are:
+
+       * ``default`` to use the default commit message provided by GitHub
+         or defined in the pull request body (see :ref:`commit message`).
+
+       * ``title+body`` means to use the title and body from the pull request
+         itself as the commit message. The pull request number will be added to
+         end of the title.
```
Comment (jd) [inline docs/source/actions/squash.rst:49]: I don't understand how you can have `message` and `commit_message` option. That does not make any sense.

**Review by jd — CHANGES_REQUESTED** [review]

File: docs/source/actions/squash.rst (hunk 2 above)
Comment (sileht) [reply docs/source/actions/squash.rst:45]: I forget to remove `message` from the other PR, this has never existed.

File: docs/source/actions/squash.rst [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+     -
+     - |premium plan tag|
+       Mergify can impersonate a GitHub user to squash a pull request.
+       If no ``bot_account`` is set, Mergify will squash the pull request
+       itself.
+
+   * - ``commit_message``
+     - string
+     - ``default``
+     - Defines what commit message to use when merging using the ``squash`` or
```
Comment (jd) [inline docs/source/actions/squash.rst:37]: Why this talks about merging since it does not merge?

File: mergify_engine/actions/squash.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            required_feature=subscription.Features.BOT_ACCOUNT,
+            missing_feature_message="Squash with `bot_account` set are disabled",
+            required_permissions=[],
+        )
+        if bot_account_result is not None:
+            return bot_account_result
+
+        if ctxt.pull["commits"] <= 1:
+            return check_api.Result(
+                check_api.Conclusion.SUCCESS, "Pull request already has one commit", ""
```
Comment (jd) [inline mergify_engine/actions/squash.py:60]:
```suggestion
                check_api.Conclusion.SUCCESS, "Pull request already is already one-commit long", ""
```

File: mergify_engine/squash_pull.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        else:
+            raise SquashFailure(
+                f"Unable to rebase: user `{bot_account}` is unknown. "
+                f"Please make sure `{bot_account}` has logged in Mergify dashboard."
+            )
+    else:
+        users = user_tokens.users
+
+    # Pick author first
+    users = sorted(users, key=lambda x: x["login"] != ctxt.pull["user"]["login"])
```
Comment (jd) [inline mergify_engine/squash_pull.py:188]: I already saw that code, might be great to have it somewhere factored?

File: mergify_engine/squash_pull.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    ]
+)
+
+
+def is_force_push_lease_reject(message: str) -> bool:
+    return (
+        "failed to push some refs" in message
+        and "[rejected]" in message
+        and "(stale info)" in message
+    )
```
Comment (jd) [inline mergify_engine/squash_pull.py:67]: This looks like duplicated code?

**Review by jd — CHANGES_REQUESTED** [review]

File: mergify_engine/squash_pull.py (hunk 6 above)
Comment (AraHaan) [reply mergify_engine/squash_pull.py:67]: Looks like that for me too.

**Comment by sileht** [comment]
Replaced by https://github.com/Mergifyio/mergify-engine/pull/2835

**Pull request closed without merging**